gprofiler/main.py

```diff
@@ -723,6 +744,45 @@ def init_pid_file(pid_file: str) -> None:
     Path(pid_file).write_text(str(os.getpid()))
 
 
+# Set limits and return path of the cgroup.
+def set_limits(cpu: float, memory: int) -> str:
+    cgroups = {}
+    logger.debug("Check if cgroup version is supported.")
+    try:
+        cgroups["cpu"] = CpuCgroup()
+        cgroups["memory"] = MemoryCgroup()
+    except UnsupportedCGroupV2:
+        logger.error("cgroup v2 is not supported by gProfiler, cpu and memory limits wouldn't be set.")
+        return
+
+    logger.debug("Prepare gProfiler cpu cgroup.")
+    try:
+        cgroups["cpu"].move_to_cgroup("gprofiler", os.getpid())
+    except AlreadyInCgroup:
+        logger.warning("gProfiler have already a cpu group.")
+
+    logger.debug("Set cpu limit in the cgroup.")
+    cgroups["cpu"].set_cpu_limit_cores(cpu)
+
+    logger.debug("Prepare gProfiler memory cgroup.")
+    try:
+        cgroups["memory"].move_to_cgroup("gprofiler", os.getpid())
+    except AlreadyInCgroup:
+        logger.warning("gProfiler have already a memory group.")
+
+    logger.debug("Set memory limit in the cgroup.")
+    cgroups["memory"].set_limit_in_bytes(memory)
+
+    return cgroups['cpu'].cgroup
```

Comment:
This API is weird. This function possibly moved the CPU cgroup, then possibly moves the memory cgroup. Moving the memory cgroup can fail, in which case you don't return the moved CPU cgroup, and the caller expects it to remain `/`.

Instead of returning a value, I think it's better to resolve the actual cgroup AFTER this function completes running.